### ISSUE TITLE
Either getPickRay or globe.pick does not work in 2D.

### ISSUE THREAD

**Issue by mramato** [AnalyticalGraphicsInc/cesium#2480]
Paste the below into Sandcastle. Clicking on the globe has no effect. Clicking off the globe alerts a failure. In 2D, it fails because `globe.pick` always returns `undefined` even when the map is picked,

``` javascript
var viewer = new Cesium.Viewer('cesiumContainer');

var handler = new Cesium.ScreenSpaceEventHandler(viewer.canvas, false);
handler.setInputAction(
    function(movement) {
        var ray = viewer.camera.getPickRay(movement.position);
        var position = viewer.scene.globe.pick(ray, viewer.scene);
        if (!Cesium.defined(position)) {
            window.alert('FAILED');
        }
    },
    Cesium.ScreenSpaceEventType.LEFT_CLICK
);

```

**Comment by SlySnake** [comment]
For 2D it seems that ray already contains the clicked position without calling globe.pick
```javascript
var viewer = new Cesium.Viewer('cesiumContainer');

var handler = new Cesium.ScreenSpaceEventHandler(viewer.canvas, false);
handler.setInputAction(
    function(movement) {
        var ray = viewer.camera.getPickRay(movement.position);
        if(viewer.camera._mode==2)
        {
            var equator_length = 2*Math.PI*6378137;
            var sideScale = equator_length/2;
            var vertScale = equator_length/4;
            console.log((ray.origin.x/sideScale*180)+","+(ray.origin.y/vertScale*90));  //lon-lat
        }
        else
        {
            var position = viewer.scene.globe.pick(ray, viewer.scene);
            if (!Cesium.defined(position)) {
            window.alert('FAILED');
            }
        }
    },
    Cesium.ScreenSpaceEventType.LEFT_CLICK
);
```
Though this might depend on HPR being (0,-90,0) while in the 2D mode.

**Comment by SlySnake** [comment]
Just noticed that the SandCastle pick demo uses pickEllipsoid rather than getPickray for all 3 projection modes. (though getPickray is used as a sub-function of pickEllipsoid.)
http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Picking.html&label=Showcases
```javascript
var cartesian = viewer.camera.pickEllipsoid(movement.endPosition, ellipsoid);
var cartographic = ellipsoid.cartesianToCartographic(cartesian);
```
pickEllipsoid calls the pick function for the current mode. Looking at pickMap2D it only cares about ray.origin from getPickRay as ray.origin is the same as the ray's destination (the x & y components.) pickMap2D then unprojects then converts from Cartographic to Cartesian. 

I noticed that pickEllipsoid handles heading changes just fine in 2DMode, but not pitch changes (only -90deg.) However I suppose pitch changes aren't officially supported in 2DMode anyhow as setView always sets pitch -90deg and roll 0 for 2DMode.

**Comment by SlySnake** [comment]
pickEllipsoid in 3D calls IntersectionTests.rayEllipsoid which picks an ellipsoid with no regard for terrain.  Globe.prototype.pick on the other hand does have regard for terrain, but 2D doesn't seem to have terrain as it's -90deg pitch only with an orthographic projection which wouldn't even notice terrain even if it was enabled.

**Comment by tlpinney** [comment]
@SlySnake 
Maybe it needs to check the projection mode and use Globe.prototype.pick for 3D and the pickEllipsoid for 2D?

I added a pull request that illustrates this at https://github.com/AnalyticalGraphicsInc/cesium/pull/2492
It is kind of hacky but I think it solves the issue with 2D and works with terrain in 3D mode.

**Comment by SlySnake** [comment]
I posted some more about this on a topic called 'camera.lookAt.range' (as Hyper Sonic.) on Cesium's forum (forum won't create a working link.) Is ImageryLayerCollection.prototype.pickImageryLayerFeatures called by viewer.scene.globe.pick? I believe the 'calling tree' is set up like this:

viewer.scene.globe.pick
-collects a list of tiles whom's sphere got intersected by the ray
-then calls GlobeSurfaceTile.prototype.pick on each of these tiles

GlobeSurfaceTile.prototype.pick
-Goes through the list of all triangles in the tile using IntersectionTests.rayTriangle
-IntersectionTests.rayTriangle calls rayTriangle 
-If rayTriangle returns a number, not only is that a hit, but is also the range of the hit

**Comment by hpinkos** [comment]
Also reported here: https://groups.google.com/d/msg/cesium-dev/PnM8vOhALfY/-OCjHEU7DgAJ

**Comment by mramato** [comment]
Mentioned again in the above linked issue.  This comes up all of the time so I marked it priority.

**Comment by xiangjg** [comment]
What can I use instead of this pickImageryLayerFeatures function in the 2D state?

**Comment by hpinkos** [comment]
Also reported by @thw0rted in #5076

**Comment by thw0rted** [comment]
Hey @pjcozzi , could you please add `category - picking`?  I missed this because it's not tagged.

**Comment by hpinkos** [comment]
Got it, thanks @thw0rted!

**Comment by krheinwald** [comment]
FYI: This bug also effects the Navigation Mixin: The Distance Legend is not showing in 2D as it uses getPickRay().

**Comment by hpinkos** [comment]
Thanks @krheinwald, but the Cesium team doesn't support a Navigation Mixin.  I would make a comment in whichever repository you're using the mixing from.

**Comment by krheinwald** [comment]
It was not my intention to get support for the Navigation Mixin. More like showing the far reaching effects of this 4yr old bug :p

**Comment by hpinkos** [comment]
3yr old bug =P
Gotcha, thanks @krheinwald! I hope we're able to address this before it becomes a 4yr old bug.
As always, if anyone has time to look into it, we're always happy to review pull requests!

**Comment by krheinwald** [comment]
Touchéz ;)

**Comment by thw0rted** [comment]
I don't have the expertise to tell, is this maybe related to #4368 ?

**Comment by cesium-concierge** [comment]
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/d/msg/cesium-dev/PnM8vOhALfY/-OCjHEU7DgAJ

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/issues/2480 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by hpinkos** [comment]
Happy to say that this bug is finally fixed! https://github.com/AnalyticalGraphicsInc/cesium/pull/6859
The fix will be included in the Cesium 1.49 release available September 3rd.

@krheinwald @thw0rted @xiangjg @WarpDrive
